data/mods/gen7/random-teams.ts

```diff
@@ -742,31 +827,47 @@ export class RandomGen7Teams extends RandomGen8Teams {
 		if (abilityData.length <= 1) return abilityData[0].name;
 
 		// Hard-code abilities here
-		if (
+		if (species.id === 'gurdurr' || (
 			abilities.has('Guts') &&
 			!abilities.has('Quick Feet') &&
-			(moves.has('facade') || (moves.has('protect') && !isDoubles) || (moves.has('sleeptalk') && moves.has('rest')))
-		) return 'Guts';
+			(moves.has('facade') || (moves.has('sleeptalk') && moves.has('rest')))
+		)) return 'Guts';
 		if (abilities.has('Moxie') && (counter.get('Physical') > 3 || moves.has('bounce')) && !isDoubles) return 'Moxie';
-		if (isDoubles) {
-			if (abilities.has('Intimidate')) return 'Intimidate';
-			if (abilities.has('Guts')) return 'Guts';
-			if (abilities.has('Storm Drain')) return 'Storm Drain';
-			if (abilities.has('Harvest')) return 'Harvest';
-			if (abilities.has('Unburden') && !abilities.has('Prankster') && !species.isMega) return 'Unburden';
-		}
 		if (species.name === 'Ambipom' && !counter.get('technician')) {
 			// If it doesn't qualify for Technician, Skill Link is useless on it
 			return 'Pickup';
 		}
+		if (
+			abilities.has('Unburden') && ['acrobatics', 'bellydrum', 'closecombat'].some(m => moves.has(m))
+		) return 'Unburden';
+		if (species.id === 'starmie') return role === 'Wallbreaker' ? 'Analytic' : 'Natural Cure';
+		if (abilities.has('Drought')) return 'Drought';
+		if (species.id === 'talonflame' && role === 'Z-Move user') return 'Gale Wings';
+		if (species.id === 'golemalola' && moves.has('return')) return 'Galvanize';
+		if (abilities.has('Gluttony') && (moves.has('recycle') || moves.has('bellydrum'))) return 'Gluttony';
+		if (abilities.has('Harvest') && (role === 'Bulky Support' || role === 'Staller')) return 'Harvest';
 		if (species.name === 'Raticate-Alola') return 'Hustle';
+		if (species.id === 'ninjask' || species.id === 'seviper') return 'Infiltrator';
+		if (species.id === 'rampardos' && role === 'Bulky Attacker') return 'Mold Breaker';
 		if (species.baseSpecies === 'Altaria') return 'Natural Cure';
+		if (species.id === 'tsareena') return 'Queenly Majesty';
+		if (role === 'AV Pivot' && abilities.has('Regenerator')) return 'Regenerator';
+		if (species.id === 'druddigon' && role === 'Bulky Support') return 'Rough Skin';
+		if (abilities.has('Scrappy') && moves.has('boomburst')) return 'Scrappy';
+		if (abilities.has('Shed Skin') && moves.has('rest') && !moves.has('sleeptalk')) return 'Shed Skin';
+		if (abilities.has('Sniper') && moves.has('focusenergy')) return 'Sniper';
+		if (species.name === 'Kommo-o' && role === 'Z-Move user') return 'Soundproof';
+		if (species.id === 'stunfisk') return 'Static';
+		if (species.id === 'breloom') return 'Technician';
+		if (abilities.has('Toxic Boost')) return 'Toxic Boost';
+		if (species.id === 'porygon2') return 'Trace';
+		if (types.has('Water') && counter.get('setup') && abilities.has('Weak Armor')) return 'Weak Armor';
```

Comment:
This also seems like something that could be broken out into top-level constants to make this easier to read.

Reply:
In this case, I disagree. There's too much variety in ability enforcement conditions and I think going through it line by line is better than trying to use multiple constants to capture them. I think this is also easier for maintenance if we ever need to make small changes to enforcement conditions.

Reply:
OK. I do think this would more readable if it were grouped by species-based checks, role-based checks, and ability-based checks. Right now they're all interspersed. Also, is there a reason the Z-move check for Kommo-o uses the species's English name instead of its ID?

Reply:
I've separated the ability enforcements into species-based and non species-based checks. Some of the species-based checks also check other things, but I think this is quite readable. While the non species-based checks aren't too long (only 8 lines), so I don't think grouping it further would be beneficial.

For the second question, no reason. I've changed Kommo-o and other species names checks in this section like Ambipom to species ID checks.